### ISSUE TITLE
Develop Swagger UI for DDCR Service Smart API

### ISSUE THREAD

**Issue by karafecho** [ResearchSoftwareInstitute/greendatatranslator#134]
The purpose of this issue is to develop a user-friendly UI to interface with the Smart API for the DDCR Service. This will allow SMEs and other users to access the DDCR integrated clinical feature tables through drop-down menus and other features, without having to execute curl commands.

**Comment by karafecho** [comment]
Swagger UI can be found at: https://ddcr.renci.org/apidocs/

**Comment by stevencox** [comment]
Looks good to me. Thanks.

**Comment by xu-hao** [comment]
Based on today's meeting with @karafecho 

- [ ] Need to add documentation.
  - [x] added more documentation.
  - [ ] potential edits from @karafecho 
- [x] Add a link to documentation in the web UI.
- [x] Delete confusing entries.
- [x] Output tabular format.

**Comment by xu-hao** [comment]
Add a link to documentation in the UI.

**Comment by xu-hao** [comment]
Delete confusing entries.

**Comment by karafecho** [comment]
@xu-hao : Please see the requests below.

1. The text/tabular view for the 'get feature association' function needs to be reformatted similar to the 'get associations to all features' function.

2. Please include the following links at the top of the API, under 'ICEES API documentation':

[patient-level tables, column headers and binning strategy](https://drive.google.com/open?id=1g-FkrlohWUv1MWbXUhclw3ZC_gsQu2PzxcwpzrnLub8),

[visit-level tables, column headers and binning strategy](https://drive.google.com/open?id=1ED0Val7kTkmAsxc26TqhvxlM7wsCzh7c_NnJXiiNMJw)

[ICEES reference slide deck](https://drive.google.com/open?id=1KvlbmQH3MPbtoFiHG7Ia2z1UbR4lw_xH)

**CORRECTION:** Instead of links to individual files, please add a link to the [ICEES folder](https://drive.google.com/open?id=12TgOZMFkWQLMhjZeN4RVzdxvlt1VYcO8).

3. Reorder functionalities as follows: (1) 'get features', (2) 'get feature association', (3) 'get associations to all features'

4. For consistency, please make the following word changes to the description of each functionality:

a. Replace 'Create a new cohort by a set of feature variables and set the cohort id. Even if a cohort has already been created for this set before, a new cohort is created.' with 'Cohort discovery. Users define a cohort using any number of defined feature variables as input parameters, and the service returns a sample size. A new cohort is created even if a cohort was previously created using the same input parameters.'

b. Replace 'get features' with 'Feature-rich cohort discovery: users select a predefined cohort as the input parameter, and the service returns a profile of that cohort in terms of all feature variables.

c. Replace 'get feature association' with 'Hypothesis-driven 2 x 2 feature associations: users select a predefined cohort and two feature variables, and the service returns a 2 x 2 feature table with a corresponding Chi Square statistic and P value.'

d. Replace 'get associations to all features' with 'Exploratory 1 X N feature associations: users select a predefined cohort and a feature variable of interest, and the service returns a 1 x N feature table with corrected Chi Square statistics and associated P values.'

5. Is it possible to move the 'response content type' to the middle column? It's easy to overlook in the current position.

6. Can we include documentation or a dictionary for versioning of tables?

If you'd prefer that I complete the above tasks, just let me know.

@xu-hao : Note that I made a correction to item 2.

@xu-hao : In terms of prioritization, I'd consider items 2, 3, and 4 plus the acknowledgement text below to be the most important.

**Comment by karafecho** [comment]
@xu-hao : One more thing: please add the text below beneath the text for the terms and conditions of the service. Thanks!

"We kindly request that Translator team members provide proper attribution for any products (e.g., manuscripts, podium presentations, software) derived from work related to Green Team's clinical datasets. Attribution should include acknowledgement of the funder (National Center for Advancing Translational Sciences [NCATS], Biomedical Data Translator Program awards, OT3TR002020 and OT2TR002514), the North Carolina Translational and Clinical Sciences (NC TraCS) Institute (NCATS, Center for Translational Science Award, UL1TR002489), UNC Hospitals and Health Care System, and all Green Team members who contributed to the work."

@xu-hao : I added an additional grant number to the text above.
@xu-hao @ColinKCurtis : As discussed today, the addition of the above text is high priority and should be completed prior to the hackathon. Thanks!

**Comment by karafecho** [comment]
@xu-hao : I believe you used panda.qcut for binning. I'm working with the data in SAS JMP and finding that the quintiles do not appear to align with yours. (See screenshots below for example distributions.) I think you binned by 'values' whereas I binned by 'population distribution'. I'm not sure which approach is best, but I do want to better understand the binning strategy that you used.

![image](https://user-images.githubusercontent.com/26254388/45170913-0977e880-b1cf-11e8-90c9-5f19836245d0.png)

**Comment by xu-hao** [comment]
@karafecho 
It seems that the data I get from qcut matches the data on your screenshot:

For example, for `EstResidentialDensity`:
```
Categories (5, interval[float64]): [(-0.001, 1250.0] < (1250.0, 1688.0] < (1688.0, 2204.0] <
                                    (2204.0, 2662.0] < (2662.0, 9653.0]]
```
matches the `Quantiles` tab in your screenshot
```
count    23093.000000
mean      2124.561815
std       1146.372076
min          0.000000
25%       1387.000000
50%       1987.000000
75%       2545.000000
max       9653.000000
```
matches the `Summary Statistics` tab in your screenshot

Numbers of samples in each category is not shown directly on the screenshot, so I cannot compare with the result of pandas. If there is a mismatch, one possible cause of mismatch could be that SAS JMP and pandas handles values on the boundary differently. If you look at the output of pandas, all the categories are left exclusive and right inclusive. If SAS JMP has a different algorithm, then it will give different numbers.

**Comment by karafecho** [comment]
Thanks, Hao. I'll look into this issue.

Note that I changed the links in item 2 above and added a second grant number and revised text to the proposed citation.

**Comment by karafecho** [comment]
For 'get definition of a cohort': I think we should return the sample size with the features that define a cohort.

**Comment by xu-hao** [comment]
* [ ] The text/tabular view for the 'get feature association' function needs to be reformatted similar to the 'get associations to all features' function.
 * [x] include the following links at the top of the API, under 'ICEES API documentation': add a link to the ICEES folder.
 * [x] Reorder functionalities as follows: (1) 'get features', (2) 'get feature association', (3) 'get associations to all features'
*  [x] For consistency, please make the following word changes to the description of each functionality:
* [ ] move the 'response content type' to the middle column? It's easy to overlook in the current position.
* [x] Can we include documentation or a dictionary for versioning of tables?
* [x] add the text below beneath the text for the terms and conditions of the service.
*  [x] For 'get definition of a cohort': I think we should return the sample size with the features that define a cohort.

**Comment by xu-hao** [comment]
@ColinKCurtis : Could you please take at look at the following feature request?

3. Reorder functionalities as follows: (1) 'get features', (2) 'get feature association', (3) 'get associations to all features'

**Comment by xu-hao** [comment]
@karafecho 
1. The text/tabular view for the 'get feature association' function needs to be reformatted similar to the 'get associations to all features' function.

It is already formatted similar to 'get associations to all features'. 
```
+-----------------+-----------------------+------------------------+---------+
| feature         | AgeStudyStart = 0-2   | AgeStudyStart <> 0-2   |         |
+=================+=======================+========================+=========+
| ObesityBMI = 0  | 2013     8.72%        | 21076   91.28%         | 23089   |
|                 | 100.00%  8.72%        | 99.98%  91.27%         | 99.98%  |
+-----------------+-----------------------+------------------------+---------+
| ObesityBMI <> 0 | 0      0.00%          | 4      100.00%         | 4       |
|                 | 0.00%  0.00%          | 0.02%  0.02%           | 0.02%   |
+-----------------+-----------------------+------------------------+---------+
|                 | 2013                  | 21080                  | 23093   |
|                 | 8.72%                 | 91.28%                 | 100.00% |
+-----------------+-----------------------+------------------------+---------+
+-----------+---------------+
|   p_value |   chi_squared |
+===========+===============+
|  0.943927 |       0.38204 |
+-----------+---------------+
```
Is there anything specific you want me to change?

**Comment by xu-hao** [comment]
@ColinKCurtis I got 3 working. Had to patch Flasgger. So help is no longer needed for this feature request.

**Comment by karafecho** [comment]
Thanks, Hao. Sorry for the confusion.

**Comment by karafecho** [comment]
@xu-hao @ColinKCurtis : Can we add back a link to [Hao's GitHub site](https://github.com/xu-hao/icees-api) above or below the link to my general-user documentation?

**Comment by karafecho** [comment]
FYI:

This query (see curl in second screenshot):
![image](https://user-images.githubusercontent.com/26254388/46044525-021a6f80-c0e9-11e8-8f7a-8dc030061616.png)

Yielded this result:
![image](https://user-images.githubusercontent.com/26254388/46044567-25451f00-c0e9-11e8-8fa6-8d10f54b07f3.png)

Could be an error on my part. Not sure, however.

**Comment by colinkcurtis** [comment]
> @xu-hao @ColinKCurtis : Can we add back a link to [Hao's GitHub site](https://github.com/xu-hao/icees-api) above or below the link to my general-user documentation?

I am looking at this now!

**Comment by xu-hao** [comment]
@karafecho i fixed the bug. thanks for finding it.

**Comment by karafecho** [comment]
@xu-hao @ColinKCurtis : I am wondering what the best approach is for pointing to plain-text documentation on ICEES. The information currently is available through a hyperlink from the UI to a Google folder, but I'm not sure this is the best approach. Thoughts?

**Comment by xu-hao** [comment]
@karafecho Simplest is just check them into github. Or we can check it into github and use github pages. Or we can use readthedocs. Given the current amount of documentation probably checking them into github is sufficient.

**Comment by karafecho** [comment]
@xu-hao @ColinKCurtis : I don't think we need to include any statistics (Chi Square, P values) for functionality two: feature-rich cohort discovery.

**Comment by xu-hao** [comment]
ok, i'll remove those stats.

**Comment by xu-hao** [comment]
it's done.

**Comment by karafecho** [comment]
@xu-hao @ColinKCurtis : Functionality 2, feature-rich cohort discovery, is not working for the visit-level tables, apparently because those tables do not contain the feature variable "TotalEDInpatientVisits".

![image](https://user-images.githubusercontent.com/26254388/46537939-99698a80-c880-11e8-9b2e-29985efbd41b.png)
![image](https://user-images.githubusercontent.com/26254388/46537958-a4bcb600-c880-11e8-8b55-76edfddf3f9a.png)

![image](https://user-images.githubusercontent.com/26254388/46538035-e0f01680-c880-11e8-8b9d-3d2dbac134cf.png)
![image](https://user-images.githubusercontent.com/26254388/46538053-eb121500-c880-11e8-9fbd-e2210492ad19.png)

**Comment by karafecho** [comment]
Actually, I think something generally is wrong with the visit-level tables.

![image](https://user-images.githubusercontent.com/26254388/46538575-5d372980-c882-11e8-98b6-3e1870bb46b7.png)

![image](https://user-images.githubusercontent.com/26254388/46538602-6a541880-c882-11e8-9567-9bbf40730adf.png)

![image](https://user-images.githubusercontent.com/26254388/46538690-9e2f3e00-c882-11e8-8f5e-600c564f2c25.png)

![image](https://user-images.githubusercontent.com/26254388/46538702-a6877900-c882-11e8-86fd-543bb542c47c.png)

**Comment by xu-hao** [comment]
I'll take a look at this today.

**Comment by xu-hao** [comment]
@karafecho i fixed a bug that is causing some of the errors. other errors can be resolved by changing the table from patient to visit.

**Comment by karafecho** [comment]
Oops! Well, that was a stupid mistake.

This works fine now. Thanks, Hao.

**Comment by karafecho** [comment]
I'm having some issues with the ICEES API, specifically, functionality 4. The service isn't returning all feature associations. For instance, AvgDailyPM2.5Exposure x TotalEDInpatientVisits isn't returned for year 2011 patient-level queries. I checked the underlying tables, however, and the data are definitely there. Plus, I can generate the output using functionality 3 for 2 x 2 tables.

![image](https://user-images.githubusercontent.com/26254388/47673099-117a5600-db8a-11e8-86ab-b9c44517e5db.png)
![image](https://user-images.githubusercontent.com/26254388/47673154-353d9c00-db8a-11e8-80dc-1130c248a601.png)

**Comment by karafecho** [comment]
I'm also having trouble generating results for functionality 4, year 2010, patient-level queries (see example below). This started today; I did not have trouble on Friday, except for the missing feature associations using functionality 4.

![image](https://user-images.githubusercontent.com/26254388/47673235-6b7b1b80-db8a-11e8-9e0d-73c9d1ea6eb3.png)
![image](https://user-images.githubusercontent.com/26254388/47673256-7afa6480-db8a-11e8-9c01-76eb47f4225c.png)

**Comment by xu-hao** [comment]
The reason "AvgDailyPM2.5Exposure x TotalEDInpatientVisits" is not showing is because the maximum_p_value is too low. If you increase it to 1, it will show up.

**Comment by karafecho** [comment]
Ahh! That makes perfect sense. I was confused because SAS JMP does not filter results, and I've been comparing output from the API with output from SAS JMP. I really need to keep these caveats in mind!

**Comment by karafecho** [comment]
@xu-hao : Okay, I'm confused. COHORT:22 was originally defined as {} (all patients), 2010, patient table, v 1.0.0. I know this because I used that cohort to generate the data in the ICEES MS. Plus, Colin used that cohort in the [Jupyter notebook](https://github.com/ncats/translator-workflows/blob/master/greengamma/general/ICEES_API_Notebook.ipynb). But now it is defined as follows:

{
  "return value": [
    {
      "size": 3958,
      "cohort_id": "COHORT:22",
      "features": {
        "TotalEDInpatientVisits": {
          "operator": "<",
          "value": 2
        },
        "MaxDailyPM2.5Exposure": {
          "operator": ">",
          "value": 2
        }
      }
    },

I created COHORT:56 to create a new cohort defined the same as COHORT:22 originally was, but I'm concerned because COHORT:22 is referenced in many places now. Is it possible to delete it and then redefine it? If so, we can delete COHORT:56.

**Comment by xu-hao** [comment]
Yes, you can use the PUT operator on the Flasgger UI to override cohort
definition. I've already restore COHORT:22 you can try it again.

**Comment by karafecho** [comment]
Excellent! Thanks, Hao. And yes, I totally used the wrong operator. Sorry!

**Comment by karafecho** [comment]
So, I just overwrote COHORT:22 again. Not sure why this is happening.

**Comment by karafecho** [comment]
Thanks for removing the 2011 tables from the ICEES API, Hao. As I mentioned, I think we need to refine the binning strategy for the CMAQ exposures in those tables before we openly release them.

**Comment by xu-hao** [comment]
once i get the FHIR data from James, i can try one of the neural network
based approach. what do you think?

**Comment by karafecho** [comment]
Sure, but let's chat first. I'm finding race-based differences in the PM2.5 distribution that are interesting and may have been masked had we optimized binning.

**Comment by karafecho** [comment]
@xu-hao @ColinKCurtis : WRT binning, I'm convinced that we will never identify a one-strategy-fits-all approach. Instead, I'm wondering if we should:

1. Optimize binning by precalculating tables using multiple binning strategies and incorporating a feature in the API/UI to allow users to select their preferred binning strategy.
2. Optimize binning by incorporating a feature in the API/UI to allow users to collapse contiguous bins. For example, this feature would allow users to select the preferred number of bins for TotalEDInpatientVisits, MajorRoadwayHighwayExposure, etc.

**Comment by xu-hao** [comment]
Both are doable.

**Comment by karafecho** [comment]
Great! Let's plan on this, as we discussed earlier today.

**Comment by karafecho** [comment]
Hao modified the API to include the binning optimization strategy (2) on 11/19/18. To be deployed.

**Comment by karafecho** [comment]
Steve Appold (Kenan-Flagler) sent a new data pull for the nationwide ACS data. The estimates are similar, but not identical, to the estimates that he sent previously (he used a different source file). He did not include the variable "EstProbabilityNonHispWhite" (not sure why), although he did include standard error estimates. These will be used by FHIR PIT to generate the new integrated feature tables.

**Comment by karafecho** [comment]
11/20/18: Hao deployed the new API. New operation is temporarily titled "feature_association2".

Example operation:

{
    "feature_a":{
        "AgeStudyStart":[
            {
                "operator":"=",
                "value":"0-2"
            }, {
                "operator":"between",
                "value_a":"3-17",
                "value_b":"18-34"
            }, {
                "operator":"in", 
                "values":["35-50","51-69"]
            },{
                "operator":"=",
                "value":"70+"
            }
        ]
    },
    "feature_b":{
        "ObesityBMI":[
            {
                "operator":"=",
                "value":0
            }, {
                "operator":"<>", 
                "value":0
            }
        ]
    }
}

**Comment by karafecho** [comment]
@xu-hao : I cannot get the new functionality to run. Is there something wrong with my syntax? 

{{
    "feature_a":{
        "AvgDailyPM2.5Exposure":[
            {
                "operator":"in",
                "values":[1,2]
            }, {
                "operator":"=",
                "value":3,
            }, {
                "operator":"in",
                "values":[4,5],
            }}
        ]
    },
    {"feature_b":{
        "TotalEDInpatientVisits":[
            {
                "operator":"<",
                "value":2
            }, 
        ]
    }
}}

**Comment by karafecho** [comment]
@xu-hao : When you get a chance, please replace paragraph three of the current DUA-like boilerplate with the following text:

"We kindly request that users of this service provide proper attribution for any products (e.g., manuscripts, podium presentations, software) derived from work related to ICEES. Attribution should include acknowledgement of the funder (National Center for Advancing Translational Sciences, National Institutes of Health, Biomedical Data Translator Program award [OT3TR002020 and OT2TR002514] and Center for Translational Science award [UL1TR002489], UNC Hospitals and Health Care System, and all team members who contributed to the work."

**Comment by xu-hao** [comment]
It's update.
________________________________

**Comment by karafecho** [comment]
Thanks, but I modified the wording slightly:

"We kindly request that users of this service provide proper attribution for any products (e.g., manuscripts, podium presentations, software) derived from work related to ICEES. Attribution should include acknowledgement of the funder (National Institutes of Health, National Center for Advancing Translational Sciences, Biomedical Data Translator Program [awards OT3TR002020 and OT2TR002514] and Center for Translational Science Program [award UL1TR002489]), UNC Hospitals and Health Care System, and all team members who contributed to the work."

Sorry!

**Comment by xu-hao** [comment]
It's updated.
________________________________

**Comment by karafecho** [comment]
Excellent! Thanks, Hao.
Kara

**Comment by karafecho** [comment]
New plan for exposures variables:

MaxDailyPM2.5Exposure_StudyAvg   avg of daily max pm2.5 over study period
MaxDailyOzoneExposure_StudyAvg   avg of daily max ozone over study period
AvgDailyPM2.5Exposure_StudyAvg    avg of daily avg pm2.5 over study period
AvgDailyOzoneExposure_StudyAvg    avg of daily avg ozone over study period
MaxDailyPM2.5Exposure_StudyMax   max of daily max pm2.5 over study period
MaxDailyOzoneExposure_StudyMax   max of daily max ozone over study period
AvgDailyPM2.5Exposure_StudyMax   max of daily avg pm2.5 over study period
AvgDailyOzoneExposure_StudyMax   max of daily avg ozone over study period

Use existing variables as default variables and map them to the newly added variables as follows:

MaxDailyPM2.5Exposure -> MaxDailyPM2.5Exposure_StudyMax
MaxDailyOzoneExposure -> MaxDailyOzoneExposure_StudyMax
AvgDailyPM2.5Exposure -> AvgDailyPM2.5Exposure_StudyAvg
AvgDailyOzoneExposure -> AvgDailyOzoneExposure_StudyAvg

**Comment by karafecho** [comment]
@xu-hao @ColinKCurtis : FYI.


  | EVENT (>=2 visits) | No EVENT (<2 visits)
-- | -- | --
Tx | a | b
No Tx | c | d
  |   |  
RR = probability of event in Tx group/probability of event in no   Tx group = (a/(a+b))/(c(c+d))
  |   |  
OR = odds of event in Tx group/odds of event in no Tx group =   (a/b)/(c/d) = ad/bc

**Comment by karafecho** [comment]
In ICEES the table would equate to the following:

  | NO EVENT | EVENT
-- | -- | --
No Tx | d | c
Tx | b | a
